File: FoosGame.m
classdef FoosGame
    properties 
        NUM_FOOSPLAYERS = 26;
        NUM_FIELDED = 22;
        ITER_PER_QUARTER = 200;
    
        MAX_MESSAGE_LENGTH = 1024;
        DELIMITER = ' ';

        game_state = [];
        comm = ''
    end
    methods
        function obj = FoosGame(GAMEID)
            obj.comm = tcpip('crisco.seas.harvard.edu', 8080);
            fopen(obj.comm);

            game_i = num2str(GAMEID); 
            fwrite(obj.comm, game_i);

            while obj.comm.BytesAvailable == 0 end
            data = fread(obj.comm, obj.comm.BytesAvailable, 'char');
            game_id = sprintf('%s', data);
            
            fprintf(['Waiting for game ' game_id '\n'])

            while obj.comm.BytesAvailable == 0 end
            data = fread(obj.comm, obj.comm.BytesAvailable, 'char');
            ready_msg = sprintf('%s', data);

            fprintf([ready_msg '\n'])
            obj.game_state = [0; 0; 0; 0];
        end
        function delete(obj)
            fclose(obj.comm)
        end
        
        % -------------------------------------
        function valid = is_valid(obj, move)
            % Makes sense as a roster
            if length(move) ~= obj.NUM_FOOSPLAYERS
                valid = false;
                return;
            end

            % Has exactly NUM_FIELDED players on the field
            if sum((abs(move)<5)) ~= obj.NUM_FIELDED
                valid = false;
                return;
            end
            
            % If this is a new quarter, then any fielded roster is ok
            if mod(obj.game_state(3), obj.ITER_PER_QUARTER) == 0
                valid = true;
                return;
            end

            % Else we're mid-quarter and can only move one player to an adjacent row
            player_pos = obj.game_state(5:(5+obj.NUM_FOOSPLAYERS));
            valid = (sum(abs(player_pos-move)) <= 1);
           return;
        end
        
        % --------------------------------------------------
        function game_state = make_move(obj, move)
            %Check if this is a valid move and display a warning
            if ~obj.is_valid(move)
                fprintf('**WARNING!! SUBMITTING INVALID MOVE**\n')
            end

            move_string = mat2str(move);
            move_string = move_string(2:end-1);
            fprintf(['Sending: ' move_string '\n']);
            
            fwrite(obj.comm, move_string);
            while obj.comm.BytesAvailable == 0 end
            data = fread(obj.comm, obj.comm.BytesAvailable, 'char');
            message = sprintf('%s', data);

            fprintf(['Recieving: ' message '\n']);
            if ~data
                fprintf('GAME KILLED\n')
                exit
            end

            game_state = str2num(message);
        end
    
    end   % END OF METHODS
    
    
end

File: Challenge2013/GameSource/FoosGame.py
#!/usr/bin/env python
"""
A simple foos client
"""
import socket

NUM_FOOSPLAYERS = 26
NUM_FIELDED = 22
ITER_PER_QUARTER = 200

MAX_MESSAGE_LENGTH = 1024
DELIMITER = ' '

class FoosGame:
  def __init__(self, game_id):
    self.comm = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    self.comm.connect(('crisco.seas.harvard.edu',8080))
    #self.comm.connect(('localhost',8080))
    self.comm.send(game_id)
    data = self.comm.recv(MAX_MESSAGE_LENGTH)
    print "Waiting for game", data
    data = self.comm.recv(MAX_MESSAGE_LENGTH)
    print data
    assert data == "READY"
    self.game_state = [0, 0, 0, 0]

  def __del__(self):
    self.comm.close()


  def is_valid(self, move):
    # Makes sense as a roster
    if len(move) != NUM_FOOSPLAYERS:
      return False

    # Has exactly NUM_FIELDED players on the field
    if sum([abs(p) < 5 for p in move]) != NUM_FIELDED:
      return False

    # If this is a new quarter, then any fielded roster is ok
    if self.game_state[2] % ITER_PER_QUARTER == 0:
      return True

    # Else we're mid-quarter and can only move one player to an adjacent row
    player_pos = self.game_state[4:(4+NUM_FOOSPLAYERS)]
    is_one_moved = (sum([abs(s-p) for (s,p) in zip(player_pos,move)]) <= 1)
    return is_one_moved


  def make_move(self, move):
    # Check if this is a valid move and display a warning
    if not self.is_valid(move):
      print "**WARNING!! SUBMITTING INVALID MOVE**"
    move_string = DELIMITER.join(map(str,move))
    print "Sending: ", move_string
    self.comm.send(move_string)
    message = self.comm.recv(MAX_MESSAGE_LENGTH)
    print "Recieving: ", message
    if not message:
      print "GAME KILLED"
      exit()
    self.game_state = [int(x) for x in message.split(DELIMITER)]
    return self.game_state

